src/parser/jobs/sam/modules/Shoha.tsx

```diff
@@ -0,0 +1,272 @@
+
+import STATUSES from 'data/STATUSES'
+
+import {ActionLink} from 'components/ui/DbLink'
+import TimeLineChart from 'components/ui/TimeLineChart'
+import ACTIONS from 'data/ACTIONS'
+import JOBS from 'data/JOBS'
+import {CastEvent, Event} from 'fflogs'
+import Module, {dependency, DISPLAY_MODE} from 'parser/core/Module'
+import Checklist, {Requirement, Rule} from 'parser/core/modules/Checklist'
+import Combatants from 'parser/core/modules/Combatants'
+import {ComboEvent} from 'parser/core/modules/Combos'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+
+const GENERATORS = {
+	[ACTIONS.HIGANBANA.id]: 1,
+	[ACTIONS.KAESHI_HIGANBANA.id]: 1,
+	[ACTIONS.TENKA_GOKEN.id]: 1,
+	[ACTIONS.KAESHI_GOKEN.id]: 1,
+	[ACTIONS.MIDARE_SETSUGEKKA.id]: 1,
+	[ACTIONS.KAESHI_SETSUGEKKA.id]: 1,
+}
+
+const SPENDERS = {
+	[ACTIONS.SHOHA.id]: 3,
+}
+
+const MAX_STACKS = 3
+
+class StackState {
+	t?: number
+	y?: number
+}
+
+export default class Shoha extends Module {
+	static handle = 'shoha'
+	static title = t('sam.shoha.title')`Meditation Timeline`
+	static displayMode = DISPLAY_MODE.FULL
+
+	private stacks = 0
+	private shohaUses = 0
+	private stackHistory: StackState[] = []
+	private wasteBySource = {
+		[ACTIONS.HIGANBANA.id]: 0,
+		[ACTIONS.KAESHI_HIGANBANA.id]: 0,
+		[ACTIONS.TENKA_GOKEN.id]: 0,
+		[ACTIONS.KAESHI_GOKEN.id]: 0,
+		[ACTIONS.MIDARE_SETSUGEKKA.id]: 0,
+		[ACTIONS.KAESHI_SETSUGEKKA.id]: 0,
+		[ACTIONS.RAISE.id]: 0,
+	}
+	private leftoverStacks = 0
+	private totalGeneratedStacks = 0 // Keep track of the total amount of generated stacks over the fight
+
+	@dependency private checklist!: Checklist
+	@dependency private suggestions!: Suggestions
+	@dependency private combatants!: Combatants
+
+	protected init() {
+		this.addHook('init', this.pushToHistory)
+		this.addHook(
+			'cast',
+			{
+				by: 'player',
+				abilityId: Object.keys(GENERATORS).map(Number),
+			},
+			this.onGenerator,
+		)
+		this.addHook(
+			'cast',
+			{
+				by: 'player',
+				abilityId: Object.keys(SPENDERS).map(Number),
+			},
+			this.onSpender,
+		)
+		this.addEventHook([
+			'applybuff', 'applybuffstack'],
+			{
+			to: 'player',
+			abilityId: STATUSES.MEDITATION.id,
+			}, this.onMeditateGain)
+
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onGenerator(event: CastEvent) {
+		if (this.stacks === MAX_STACKS) {
+		const abilityId = event.ability.guid
+		const generatedStacks = GENERATORS[abilityId]
+
+		this.addGeneratedStackAndPush(generatedStacks, abilityId)
+	}
+	}
+
+		// This is exclusively for stacks gained via Meditate channel. and not any stacks gained from GCD casts
+	private onMeditateGain() {
+			this.stacks++
+			this.totalGeneratedStacks++
+
+			// just a safety net
+			if (this.stacks > MAX_STACKS) {
+				this.stacks = MAX_STACKS
+			}
+
+			this.pushToHistory()
+	}
+
+	private addGeneratedStackAndPush(generatedStacks: number, abilityId: number) {
```

Comment:
You sorta dupe this in Meditate anyway, but then only really use it once. Any reason not to just roll it into the normal generator method? I don't think it makes the function particularly unweildy.

Reply:
I really should rename that to AddWaste or something. It started out as the go to adding stacks method, but now I've converted it to be used as a function to handle waste from the generators used at cap. 

I could probably just merge this into the Iaijustu cast code since the whole point of that code is just to check if we're already capped and then move into this function to add it to waste

Reply:
Eh, no issue with functions that are called once, esp. if they help split logic into more approachable blocks. Accurate names are 👌 though.